---
title: "Gene Expression"
author: "ZDR"
date: 2018-09-03
output:
  github_document:
    toc: true
urlcolor: blue
---

## Setup
<!-- -------------------------------------------------- -->

The following is a look at the UCI machine learning repository
[https://archive.ics.uci.edu/ml/datasets/gene+expression+cancer+RNA-Seq](gene
expression) database. These data have dimensionality on the order of `20,000`.
We will apply randomized dimension reduction -- informed by the
Johnson-Lindenstrauss lemma -- to make subsequent analysis more tractable.

```{r setup, echo=FALSE}
library(tidyverse)
library(matrixStats)

## Data parameters
file_data <- "~/Sync/data/gene_expression/data.csv"

## Script parameters
eps <- 0.1

## Helper functions

## row scale function, https://www.r-bloggers.com/a-faster-scale-function/
row_scale = function(x,
    center = TRUE,
    scale = TRUE,
    add_attr = TRUE,
    rows = NULL,
    cols = NULL) {

    if (!is.null(rows) && !is.null(cols)) {
        x <- x[rows, cols, drop = FALSE]
    } else if (!is.null(rows)) {
        x <- x[rows, , drop = FALSE]
    } else if (!is.null(cols)) {
        x <- x[, cols, drop = FALSE]
    }

  # Get the column means
  ################
    cm = rowMeans(x, na.rm = TRUE)
  # Get the column sd
  ################
    if (scale) {
        csd = rowSds(x, center = cm)
    } else {
        # just divide by 1 if not
        csd = rep(1, length = length(cm))
    }
    if (!center) {
        # just subtract 0
        cm = rep(0, length = length(cm))
    }
    x = (x - cm) / csd
    if (add_attr) {
        if (center) {
            attr(x, "scaled:center") <- cm
        }
        if (scale) {
            attr(x, "scaled:scale") <- csd
        }
    }
    return(x)
}
```

First we need to read and wrangle the data. Let's standardize every variable
before further analysis -- off-center data will play poorly with projection, and
high-value data could lead to overflow.

```{r read_wrangle}
df_raw <-
    file_data %>%
    read_csv()

mat_data <-
    df_raw %>%
    select(-X1) %>%
    as.matrix(.) %>%
    row_scale(.)
```

Let's check out the raw size of the original data.

```{r raw_size}
dim(mat_data)
```

We have woefully few observations (`r dim(mat_data)[1]`) for the large number of
variables (`r dim(mat_data)[2]`) in these data. Let's see if we find a
reasonable dimension reduction.

## J-L Application
<!-- -------------------------------------------------- -->

First, let's carry out a Johnson-Lindenstrauss informed projection. A
high-dimensional gaussian is nearly orthogonal, so we need only draw from a
multivariate gaussian of the appropriate size and normalize.

```{r jl_dim}
## Make reproducible
set.seed(101)
## Calculate dimension
C <- 8                             # Over-samping factor
n <- dim(mat_data)[1]              # Observations
d <- dim(mat_data)[2]              # Dimensionality

k <- C * ceiling(log(n) / eps ^ 2) # J-L dimension
## Random `projection`
P <-
    map_dfc(1:k, function(a) {x = rnorm(d); x / norm(x, type = "2")}) %>%
    as.matrix(.)
## Project the data
mat_proj <-
    mat_data %*% P
## Match the original average distance; Subsample for speed
Ind <- sample(x = 1:n, size = 500)

D_sub_orig <-
    mat_data[Ind, ] %>%
    dist()

D_sub_fix <-
    mat_proj[Ind, ] %>%
    dist()

factor <-
    mean(D_sub_orig, na.rm = TRUE) / mean(D_sub_fix, na.rm = TRUE)

mat_proj <-
    mat_proj * factor
```

Randomly sub-select the distances and check they are close.

```{r jl_sub_check}
## Re-compute distances with proper scaling
Ind <- sample(x = 1:n, size = 500) # New draw

D_sub_orig <-
    mat_data[Ind, ] %>%
    dist()

D_sub_proj <-
    mat_proj[Ind, ] %>%
    dist()

## Compute quantiles of discrepancy
R_diff <- (D_sub_proj - D_sub_orig) / D_sub_orig
qt <- quantile(R_diff)
qt
```

Our tolerance was `r eps`, and the extremes of the distance discrepancy were `r qt[1]`
and `r qt[5]`. In this case, we found the projection in one shot; in practice, we
might have to re-draw until an acceptable projection is found. J-L guarantees we
can do this in polynomial time.
